### PR TITLE
Support for automatic type coercer in hive parquet tables

### PR DESCRIPTION
**Pull request by krvikash**
## Description
Support for automatic type coercer in hive parquet tables

-  bigint to double
- timestamp to string/varchar

## Release notes

(X) Release notes are required. Please propose a release note for me.

```markdown
# Hive
* Add support for reading integers in parquet files as DOUBLE columns in Trino. ({issue}`21509`)
* Add support for reading timestamps in parquet files as VARCHAR columns in Trino. ({issue}`21509`)
```

### CODE REVIEW

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSource.java
Comment (Praveen2112) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSource.java:1]: Typo in `Parquet` file format name.

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveCoercionOnUnpartitionedTable.java [hunk 1]
```diff
@@ -183,6 +198,13 @@ public void testHiveCoercionWithDifferentTimestampPrecision()
         doTestHiveCoercionWithDifferentTimestampPrecision(HIVE_TIMESTAMP_COERCION_ORC);
     }
 
+    @Requires(ParquetRequirements.class)
```
Comment (Praveen2112) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveCoercionOnUnpartitionedTable.java:201]: Can we enable the `testHiveCoercionWithDifferentTimestampPrecision()` for parquet format ?

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetTypeTranslator.java [hunk 2]
```diff
@@ -35,6 +39,11 @@ private ParquetTypeTranslator() {}
                 return Optional.of(new IntegerNumberToDoubleCoercer<>(BIGINT));
             }
         }
+        if (toTrinoType instanceof VarcharType varcharType) {
+            if (fromParquetType == INT96) {
+                return Optional.of(new TimestampCoercer.LongTimestampToVarcharCoercer(TIMESTAMP_NANOS, varcharType));
```
Comment (Praveen2112) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetTypeTranslator.java:44]: nit: static import for `LongTimestampToVarcharCoercer`

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveCoercionOnUnpartitionedTable.java [hunk 3]
```diff
@@ -219,6 +253,133 @@ protected Map<ColumnContext, String> expectedExceptionsWithTrinoContext()
                 .put(columnContext("orc", "timestamp_row_to_row"), "Cannot read SQL type 'varchar' from ORC stream '.timestamp_row_to_row.timestamp2string' of type TIMESTAMP with attributes {}")
                 .put(columnContext("orc", "timestamp_list_to_list"), "Cannot read SQL type 'varchar' from ORC stream '.timestamp_row_to_row.timestamp2string' of type TIMESTAMP with attributes {}")
                 .put(columnContext("orc", "timestamp_map_to_map"), "Cannot read SQL type 'varchar' from ORC stream '.timestamp_row_to_row.timestamp2string' of type TIMESTAMP with attributes {}")
+                // PARQUET
+                .put(columnContext("parquet", "row_to_row"), "Unsupported Trino column type (varchar) for Parquet column ([row_to_row, bi2vc] optional int64 bi2vc (INTEGER(64,true)))")
```
Comment (marcinsbd) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveCoercionOnUnpartitionedTable.java:257]: is there a way to generate complete list of possible conversions (fromType_to_toType) that will be used to verify automatically that we cover all test cases of possible conversions?

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/parquet/TestDecimalCoercers.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+import static io.trino.spi.type.DecimalType.createDecimalType;
+import static io.trino.spi.type.IntegerType.INTEGER;
+import static io.trino.spi.type.SmallintType.SMALLINT;
+import static io.trino.spi.type.TinyintType.TINYINT;
+import static org.apache.parquet.schema.PrimitiveType.PrimitiveTypeName;
+import static org.apache.parquet.schema.PrimitiveType.PrimitiveTypeName.INT32;
+import static org.apache.parquet.schema.PrimitiveType.PrimitiveTypeName.INT64;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestDecimalCoercers
```
Comment (Praveen2112) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/parquet/TestDecimalCoercers.java:36]: Is the coerced value is different from ORC ? In that case reuse the test cases ?

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetTypeTranslator.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+public final class ParquetTypeTranslator
+{
+    private ParquetTypeTranslator() {}
+
+    public static Optional<TypeCoercer<? extends Type, ? extends Type>> createCoercer(PrimitiveType fromParquetType, Type toTrinoType)
+    {
+        PrimitiveTypeName fromParquetTypeName = fromParquetType.getPrimitiveTypeName();
+
+        if (toTrinoType instanceof DoubleType) {
+            if (fromParquetTypeName == INT64) {
```
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetTypeTranslator.java:49]: Yes, it is a problem in the orc coercion as well

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java [hunk 6]
```diff
@@ -435,13 +439,30 @@ public static ConnectorPageSource createParquetPageSource(
...
+            Optional<TypeCoercer<?, ?>> coercer = Optional.empty();
+            ColumnIO columnIO = lookupColumnByName(messageColumn, columnName);
+            if (columnIO != null && columnIO.getType().isPrimitive()) {
+                PrimitiveType primitiveType = columnIO.getType().asPrimitiveType();
+                coercer = createCoercer(primitiveType.getPrimitiveTypeName(), primitiveType.getLogicalTypeAnnotation(), column.getBaseType());
+            }
+
+            io.trino.spi.type.Type readType = column.getBaseType();
+            if (coercer.isPresent()) {
+                readType = coercer.get().getFromType();
```
Comment (Praveen2112) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:452]: Can we use `coercer.map(::getFromType).orElse(column::getBaseType)`

**Review by Praveen2112 — APPROVED** [review]
I would wait for @raunaqmorarka review here.

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/coercions/TestIntegerNumberToDoubleCoercerParquet.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+import org.apache.parquet.schema.PrimitiveType;
+import org.apache.parquet.schema.Types;
+
+import static io.trino.spi.predicate.Utils.nativeValueToBlock;
+import static io.trino.spi.type.DoubleType.DOUBLE;
+import static org.apache.parquet.schema.LogicalTypeAnnotation.IntLogicalTypeAnnotation.intType;
+import static org.apache.parquet.schema.PrimitiveType.PrimitiveTypeName.INT32;
+import static org.apache.parquet.schema.PrimitiveType.PrimitiveTypeName.INT64;
+
+public class TestIntegerNumberToDoubleCoercerParquet
```
Comment (raunaqmorarka) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/coercions/TestIntegerNumberToDoubleCoercerParquet.java:32]: I don't see the point of running this test separately for parquet. The coercion logic is mostly outside of parquet, the only part in parquet is the selection of TypeCoercer based on parquet and Trino types and the page source factory change. We should just have a query runner test which creates parquet file with tinyint/smallint/int/bigint types and then reads that back as a table with all double types. If that's redundant with the product tests we already have, then we can skip it

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/coercions/TestTimestampCoercerParquet.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+import io.trino.spi.block.Block;
+import io.trino.spi.type.Type;
+import org.junit.jupiter.api.Test;
+
+import static io.trino.plugin.hive.parquet.ParquetTypeTranslator.createCoercer;
+import static io.trino.spi.predicate.Utils.nativeValueToBlock;
+import static org.apache.parquet.schema.PrimitiveType.PrimitiveTypeName.INT96;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public class TestTimestampCoercerParquet
```
Comment (raunaqmorarka) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/coercions/TestTimestampCoercerParquet.java:26]: Same comment as TestIntegerNumberToDoubleCoercerParquet

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/IntegerNumberToDoubleCoercer.java [hunk 9]
```diff
@@ -32,6 +37,17 @@ public IntegerNumberToDoubleCoercer(F fromType)
     @Override
     protected void applyCoercedValue(BlockBuilder blockBuilder, Block block, int position)
     {
+        long value = fromType.getLong(block, position);
+        // IEEE 754 double-precision can guarantee this for up to 53 bits (52 bits of significand + the implicit leading 1 bit)
+        // https://stackoverflow.com/questions/43655668/are-all-integer-values-perfectly-represented-as-doubles
+        if (overflow(value)) {
+            throw new TrinoException(INVALID_CAST_ARGUMENT, "Cannot read parquet value '%s' as DOUBLE".formatted(value));
```
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/IntegerNumberToDoubleCoercer.java]: I would consider that a bug in Hive, ideally implicit coercions shouldn't change the value as the user hasn't specified an explicit CAST. You could try raising a JIRA in Apache Hive project to see if they have a different opinion.

**Review by raunaqmorarka — APPROVED** [review]

**Review by marcinsbd — COMMENTED** [review]
LGTM

**Comment by findepi** [comment]
https://github.com/trinodb/trino/issues/21975 seems related to this PR

**Pull request merged**